File: backend/src/models/Flight.ts
import { Schema, model, Document } from 'mongoose';
import { FlightStatus, BarcodeData, AirportLocation } from '@my-app/shared';

// Extend the shared Flight interface for MongoDB Document
export interface IFlight extends Document {
  userId: Schema.Types.ObjectId;
  
  // Basic flight info
  airline: string;
  airlineCode?: string;
  airlineLogo?: string;
  flightNumber: string;
  confirmationCode: string;
  eticketNumber?: string;
  
  // Route info
  origin: AirportLocation;
  destination: AirportLocation;
  distance?: number; // in miles
  duration?: number; // in minutes
  
  // Time info
  scheduledDepartureTime: Date;
  scheduledArrivalTime: Date;
  actualDepartureTime?: Date;
  actualArrivalTime?: Date;
  boardingTime?: Date;
  
  // Boarding details
  seatNumber: string;
  boardingGroup?: string;
  boardingZone?: string;
  sequenceNumber?: string;
  classOfService?: 'economy' | 'premium-economy' | 'business' | 'first';
  
  // Flight stats
  points?: number; // loyalty points earned
  
  // Documents
  boardingPassUrl?: string; // S3 URL
  barcode?: BarcodeData;
  
  // Status
  status: FlightStatus;
  notes?: string;
  
  createdAt: Date;
  updatedAt: Date;
  
  // Methods
  calculatePoints(): number;
}

// Airport location sub-schema
const airportLocationSchema = new Schema({
  airportCode: {
    type: String,
    required: true,
    uppercase: true,
    minlength: 3,
    maxlength: 3
  },
  airportName: String,
  city: {
    type: String,
    required: true
  },
  country: {
    type: String,
    required: false
  },
  terminal: String,
  gate: String
}, { _id: false });

// Barcode sub-schema
const barcodeSchema = new Schema({
  type: {
    type: String,
    enum: ['PDF417', 'QR', 'AZTEC', 'OTHER'],
    required: true
  },
  value: {
    type: String,
    required: true
  }
}, { _id: false });

const flightSchema = new Schema<IFlight>({
  userId: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  
  // Basic flight info
  airline: {
    type: String,
    required: false,
    default: 'Unknown'
  },
  airlineCode: {
    type: String,
    enum: ['AA', 'DL', 'UA', 'WN', 'NK', 'F9', 'B6', 'AS', 'HA', 'OTHER']
  },
  airlineLogo: String,
  flightNumber: {
    type: String,
    required: false,
    default: 'N/A'
  },
  confirmationCode: {
    type: String,
    required: false,
    default: 'MANUAL'
  },
  eticketNumber: String,
  
  // Route info
  origin: {
    type: airportLocationSchema,
    required: true
  },
  destination: {
    type: airportLocationSchema,
    required: true
  },
  distance: Number, // miles
  duration: Number, // minutes
  
  // Time info
  scheduledDepartureTime: {
    type: Date,
    required: true,
    index: true
  },
  scheduledArrivalTime: {
    type: Date,
    required: true
  },
  actualDepartureTime: Date,
  actualArrivalTime: Date,
  boardingTime: Date,
  
  // Boarding details
  seatNumber: {
    type: String,
    required: false,
    default: 'N/A'
  },
  boardingGroup: String,
  boardingZone: String,
  sequenceNumber: String,
  classOfService: {
    type: String,
    enum: ['economy', 'premium-economy', 'business', 'first']
  },
  
  // Flight stats
  points: Number,
  
  // Documents
  boardingPassUrl: String,
  barcode: barcodeSchema,
  
  // Status
  status: {
    type: String,
    enum: ['upcoming', 'completed', 'cancelled', 'delayed', 'in-flight'],
    default: 'upcoming',
    required: true,
    index: true
  },
  notes: String
}, {
  timestamps: true
});

// Indexes for common queries
flightSchema.index({ userId: 1, scheduledDepartureTime: -1 });
flightSchema.index({ userId: 1, status: 1 });
flightSchema.index({ userId: 1, airline: 1 });
flightSchema.index({ 'origin.airportCode': 1, 'destination.airportCode': 1 });

// Calculate points based on distance and class
flightSchema.methods.calculatePoints = function(): number {
  if (!this.distance) return 0;
  
  let basePoints = Math.floor(this.distance);
  
  // Apply class multipliers
  switch (this.classOfService) {
    case 'first':
      basePoints *= 3;
      break;
    case 'business':
      basePoints *= 2;
      break;
    case 'premium-economy':
      basePoints *= 1.5;
      break;
    default:
      // economy stays at 1x
  }
  
  return Math.floor(basePoints);
};

// Pre-save middleware to calculate points
flightSchema.pre('save', function(next) {
  if (this.isModified('distance') || this.isModified('classOfService')) {
    this.points = this.calculatePoints();
  }
  next();
});

export const Flight = model<IFlight>('Flight', flightSchema);
export default Flight;

File: mobile/lib/features/auth/data/models/flight.dart
import 'package:json_annotation/json_annotation.dart';

part 'flight.g.dart';

// Flight status enum
enum FlightStatus {
  @JsonValue('upcoming')
  upcoming,
  @JsonValue('completed')
  completed,
  @JsonValue('cancelled')
  cancelled,
  @JsonValue('delayed')
  delayed,
  @JsonValue('in-flight')
  inFlight
}

// Class of service enum
enum ClassOfService {
  @JsonValue('economy')
  economy,
  @JsonValue('premium-economy')
  premiumEconomy,
  @JsonValue('business')
  business,
  @JsonValue('first')
  first
}

// Barcode type enum
enum BarcodeType {
  @JsonValue('PDF417')
  pdf417,
  @JsonValue('QR')
  qr,
  @JsonValue('AZTEC')
  aztec,
  @JsonValue('OTHER')
  other
}

@JsonSerializable()
class AirportLocation {
  final String airportCode;
  final String? airportName;
  final String city;
  final String country;
  final String? terminal;
  final String? gate;

  AirportLocation({
    required this.airportCode,
    this.airportName,
    required this.city,
    required this.country,
    this.terminal,
    this.gate,
  });

  factory AirportLocation.fromJson(Map<String, dynamic> json) =>
      _$AirportLocationFromJson(json);

  Map<String, dynamic> toJson() => _$AirportLocationToJson(this);
}

@JsonSerializable()
class BarcodeData {
  final BarcodeType type;
  final String value;

  BarcodeData({
    required this.type,
    required this.value,
  });

  factory BarcodeData.fromJson(Map<String, dynamic> json) =>
      _$BarcodeDataFromJson(json);

  Map<String, dynamic> toJson() => _$BarcodeDataToJson(this);
}

@JsonSerializable()
class Flight {
  @JsonKey(name: '_id')
  final String? id;
  final String userId;
  
  // Basic flight info
  final String airline;
  final String? airlineCode;
  final String? airlineLogo;
  final String flightNumber;
  final String confirmationCode;
  final String? eticketNumber;
  
  // Route info
  final AirportLocation origin;
  final AirportLocation destination;
  final int? distance; // in miles
  final int? duration; // in minutes
  
  // Time info
  final DateTime scheduledDepartureTime;
  final DateTime scheduledArrivalTime;
  final DateTime? actualDepartureTime;
  final DateTime? actualArrivalTime;
  final DateTime? boardingTime;
  
  // Boarding details
  final String seatNumber;
  final String? boardingGroup;
  final String? boardingZone;
  final String? sequenceNumber;
  final ClassOfService? classOfService;
  
  // Flight stats
  final int? points;
  
  // Documents
  final String? boardingPassUrl;
  final BarcodeData? barcode;
  
  // Status
  final FlightStatus status;
  final String? notes;
  
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Flight({
    this.id,
    required this.userId,
    required this.airline,
    this.airlineCode,
    this.airlineLogo,
    required this.flightNumber,
    required this.confirmationCode,
    this.eticketNumber,
    required this.origin,
    required this.destination,
    this.distance,
    this.duration,
    required this.scheduledDepartureTime,
    required this.scheduledArrivalTime,
    this.actualDepartureTime,
    this.actualArrivalTime,
    this.boardingTime,
    required this.seatNumber,
    this.boardingGroup,
    this.boardingZone,
    this.sequenceNumber,
    this.classOfService,
    this.points,
    this.boardingPassUrl,
    this.barcode,
    required this.status,
    this.notes,
    this.createdAt,
    this.updatedAt,
  });

  factory Flight.fromJson(Map<String, dynamic> json) => _$FlightFromJson(json);

  Map<String, dynamic> toJson() => _$FlightToJson(this);

  // Helper methods
  String get flightRoute => '${origin.airportCode} → ${destination.airportCode}';
  
  String get formattedDepartureTime {
    final hour = scheduledDepartureTime.hour;
    final minute = scheduledDepartureTime.minute.toString().padLeft(2, '0');
    final period = hour >= 12 ? 'PM' : 'AM';
    final displayHour = hour > 12 ? hour - 12 : (hour == 0 ? 12 : hour);
    return '$displayHour:$minute $period';
  }
  
  String get formattedArrivalTime {
    final hour = scheduledArrivalTime.hour;
    final minute = scheduledArrivalTime.minute.toString().padLeft(2, '0');
    final period = hour >= 12 ? 'PM' : 'AM';
    final displayHour = hour > 12 ? hour - 12 : (hour == 0 ? 12 : hour);
    return '$displayHour:$minute $period';
  }
  
  String get formattedDuration {
    if (duration == null) return 'N/A';
    final hours = duration! ~/ 60;
    final minutes = duration! % 60;
    return '${hours}h ${minutes}m';
  }
  
  String get statusDisplay {
    switch (status) {
      case FlightStatus.upcoming:
        return 'Upcoming';
      case FlightStatus.completed:
        return 'Completed';
      case FlightStatus.cancelled:
        return 'Cancelled';
      case FlightStatus.delayed:
        return 'Delayed';
      case FlightStatus.inFlight:
        return 'In Flight';
    }
  }
  
  String get classDisplay {
    if (classOfService == null) return 'Economy';
    switch (classOfService!) {
      case ClassOfService.economy:
        return 'Economy';
      case ClassOfService.premiumEconomy:
        return 'Premium Economy';
      case ClassOfService.business:
        return 'Business';
      case ClassOfService.first:
        return 'First';
    }
  }
}

@JsonSerializable()
class FlightStats {
  final FlightStatsSummary summary;
  final List<MonthlyFlightStats> flightsByMonth;
  final List<RouteStats> topRoutes;
  final List<AirlineStats> airlineBreakdown;
  final StatusBreakdown statusBreakdown;

  FlightStats({
    required this.summary,
    required this.flightsByMonth,
    required this.topRoutes,
    required this.airlineBreakdown,
    required this.statusBreakdown,
  });

  factory FlightStats.fromJson(Map<String, dynamic> json) =>
      _$FlightStatsFromJson(json);

  Map<String, dynamic> toJson() => _$FlightStatsToJson(this);
}

@JsonSerializable()
class FlightStatsSummary {
  final int totalFlights;
  final int totalDistance;
  final int totalDuration;
  final int totalPoints;
  final int uniqueAirlines;
  final int uniqueDestinations;
  final int uniqueCountries;

  FlightStatsSummary({
    required this.totalFlights,
    required this.totalDistance,
    required this.totalDuration,
    required this.totalPoints,
    required this.uniqueAirlines,
    required this.uniqueDestinations,
    required this.uniqueCountries,
  });

  factory FlightStatsSummary.fromJson(Map<String, dynamic> json) =>
      _$FlightStatsSummaryFromJson(json);

  Map<String, dynamic> toJson() => _$FlightStatsSummaryToJson(this);
}

@JsonSerializable()
class MonthlyFlightStats {
  final int month;
  final int year;
  final int count;
  final int distance;
  final int points;

  MonthlyFlightStats({
    required this.month,
    required this.year,
    required this.count,
    required this.distance,
    required this.points,
  });

  factory MonthlyFlightStats.fromJson(Map<String, dynamic> json) =>
      _$MonthlyFlightStatsFromJson(json);

  Map<String, dynamic> toJson() => _$MonthlyFlightStatsToJson(this);
}

@JsonSerializable()
class RouteStats {
  final String origin;
  final String destination;
  final int count;
  final int totalDistance;

  RouteStats({
    required this.origin,
    required this.destination,
    required this.count,
    required this.totalDistance,
  });

  factory RouteStats.fromJson(Map<String, dynamic> json) =>
      _$RouteStatsFromJson(json);

  Map<String, dynamic> toJson() => _$RouteStatsToJson(this);
}

@JsonSerializable()
class AirlineStats {
  final String airline;
  final int count;
  final int distance;
  final int points;

  AirlineStats({
    required this.airline,
    required this.count,
    required this.distance,
    required this.points,
  });

  factory AirlineStats.fromJson(Map<String, dynamic> json) =>
      _$AirlineStatsFromJson(json);

  Map<String, dynamic> toJson() => _$AirlineStatsToJson(this);
}

@JsonSerializable()
class StatusBreakdown {
  final int upcoming;
  final int completed;
  final int cancelled;
  final int delayed;

  StatusBreakdown({
    required this.upcoming,
    required this.completed,
    required this.cancelled,
    required this.delayed,
  });

  factory StatusBreakdown.fromJson(Map<String, dynamic> json) =>
      _$StatusBreakdownFromJson(json);

  Map<String, dynamic> toJson() => _$StatusBreakdownToJson(this);
}